src/gcp/cloudsql/permissions.ts

```diff
@@ -0,0 +1,267 @@
+import { Options } from "../../options";
+import { needProjectId, needProjectNumber } from "../../projectUtils";
+import { dataconnectP4SADomain } from "../../api";
+import { toDatabaseUser, executeSqlCmdsAsIamUser, executeSqlCmdsAsSuperUser } from "./connect";
+import { testIamPermissions } from "../iam";
+import { logger } from "../../logger";
+import { concat } from "lodash";
+import * as utils from "../../utils";
+import * as cloudSqlAdminClient from "./cloudsqladmin";
+import { FirebaseError } from "../../error";
+
+export function firebaseowner(databaseId: string) {
+  return `firebaseowner_${databaseId}_public`;
+}
+
+export function firebasereader(databaseId: string) {
+  return `firebasereader_${databaseId}_public`;
+}
+
+export function firebasewriter(databaseId: string) {
+  return `firebasewriter_${databaseId}_public`;
+}
+
+export function getDataConnectP4SA(projectNumber: string): string {
+  return `service-${projectNumber}@${dataconnectP4SADomain()}`;
+}
+
+// Returns true if "grantedRole" is granted to "granteeRole" and false otherwise.
+// Throw an error if commands fails due to another reason like connection issues.
+export async function checkRoleIsGranted(
+  options: Options,
+  instanceId: string,
+  databaseId: string,
+  grantedRole: string,
+  granteeRole: string,
+): Promise<boolean> {
+  const checkCmd = `
+    DO $$
+    DECLARE
+        role_count INTEGER;
+    BEGIN
+        -- Count the number of rows matching the criteria
+        SELECT COUNT(*)
+        INTO role_count
+        FROM
+          pg_auth_members m
+        JOIN
+          pg_roles grantee ON grantee.oid = m.member
+        JOIN
+          pg_roles granted ON granted.oid = m.roleid
+        JOIN
+          pg_roles grantor ON grantor.oid = m.grantor
+        WHERE
+          granted.rolname = '${grantedRole}'
+          AND grantee.rolname = '${granteeRole}';
+
+        -- If no rows were found, raise an exception
+        IF role_count = 0 THEN
+            RAISE EXCEPTION 'Role "%", is not granted to role "%".', '${grantedRole}', '${granteeRole}';
+        END IF;
+    END $$;
+`;
+  try {
+    await executeSqlCmdsAsIamUser(options, instanceId, databaseId, [checkCmd], true);
+    return true;
+  } catch (e) {
+    // We only return false after we confirm the error is indeed because the role isn't granted.
+    // Otherwise we propagate the error.
+    if (e instanceof FirebaseError && e.message.includes("not granted to role")) {
+      return false;
+    }
+    logger.error(`Role Check Failed: ${e}`);
+    throw e;
+  }
+}
+
+export async function iamUserIsCSQLAdmin(options: Options): Promise<boolean> {
+  const projectId = needProjectId(options);
+  const requiredPermissions = [
+    "cloudsql.databases.create",
+    "cloudsql.databases.update",
+    "cloudsql.instances.connect",
+    "cloudsql.instances.create",
+    "cloudsql.instances.get",
+    "cloudsql.instances.list",
+    "cloudsql.instances.update",
+    "cloudsql.users.create",
```

Comment:
Only `cloudsql.users.create` and `cloudsql.instances.connect` matters for CLI's use case

We already have a check in `dataconnect:sql:migrate`. We should remove `cloudsql.users.create` there, so non-admin developers can run `sql:migrate` as well. And perform admin check in application logics.

https://github.com/firebase/firebase-tools/blob/d4ac884423e9cca014e3a31e619c6b2e176caa02/src/commands/dataconnect-sql-migrate.ts#L14-L19

Reply:
Thanks, good catch. I remove `cloudsql.users.create` from the migrate  requirements. I updated the isAdmin permissions to be check for less permissions but I think `cloudsql.instances.login` and `cloudsql.instances.get` should remain.

Reply:
Technically, you can connect via sql-proxy without `cloudsql.instances.login`.

It's pretty confusing, but `login` is only needed for manual IAM authentication, not used in CLI.

https://cloud.google.com/sql/docs/postgres/iam-logins#logging-in-as-a-user

Reply:
Ok, removed login.